jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpChannelState.java

```diff
@@ -1566,18 +1566,24 @@ private void completeChannelCallback(Throwable failure)
                 stream = httpChannelState._stream;
                 assert httpChannelState._callbackFailure == null;
 
-                // Turn pending demand or unconsumed input on persistent connections into failure
+                // Turn pending demand into failure.
                 if (httpChannelState._onContentAvailable != null)
+                {
                     failure = ExceptionUtil.combine(failure, new IllegalStateException("demand pending"));
-                else if (httpChannelState.getConnectionMetaData().isPersistent())
+                }
+                // Else failure for a persistent non expecting request to not consume the request
+                else if (httpChannelState.getConnectionMetaData().isPersistent() && !httpChannelState._expects100Continue)
                     failure = ExceptionUtil.combine(failure, stream.consumeAvailable());
                 else
                 {
+                    // Else we consumeAvailable, but if the content cannot be fully consumed, then
+                    // consumeAvailable() makes the connection non-persistent and returns an exception.
+                    // This must not result in an error according to RFC2616 section 8.2.3.
                     Throwable unconsumed = stream.consumeAvailable();
```

Comment:
The connection could be persistent here, `consumeAvailable` should then be skipped.